File: CCLib/src/java/util/ArrayList.java
package java.util;

public class ArrayList<E> implements List<E> {
	private E[] array;

	public ArrayList() {
		initArray();
	}

	native private void initArray();

	//Collection
	native public boolean add(E e);
	native public void clear();
	native public boolean remove(E e);

	public int size() {
		return array.length;
	}

	//List
    native public void add(int index, E element);
    native public E remove(int index);

	public E get(int index) {
		return array[index];
	}

    public E set(int index, E element) {
    	E ret = array[index];
    	array[index] = element;
    	return ret;
    }

    public List<E> subList(int fromIndex, int toIndex) {
    	List<E> newList = new ArrayList<E>();
    	for (int i = fromIndex; i < toIndex; ++i) {
    		newList.add(get(i));
    	}
    	return newList;
    }
}

File: CCLib/src/java/lang/native/ArrayList.lua
natives["jvml.util.ArrayList"] = natives["jvml.util.ArrayList"] or {}

natives["jvml.util.ArrayList"]["initArray()V"] = function(this)
	-- array must be initialized natively, as generics can't be used for array construction
	local arr = newArray(getArrayClass("[Ljava.lang.Object;"), 0)
	setObjectField(this, "array", arr)
end

natives["jvml.util.ArrayList"]["add(Ljava/lang/Object;)Z"] = function(this, e)
	local arr = getObjectField(this, "array")
	arr[5][arr[4] + 1] = e
	arr[4] = arr[4] + 1
	return 1 -- java booleans are 1/0
end

natives["jvml.util.ArrayList"]["add(ILjava/lang/Object;)Z"] = function(this, i, e)
	local arr = getObjectField(this, "array")
	assert(i < arr[4] and i >= 0, "ArrayList index out of bounds")
	table.insert(arr[5], i + 1, e)
	arr[4] = arr[4] + 1
	return 1
end

natives["jvml.util.ArrayList"]["clear()V"] = function(this)
	local arr = getObjectField(this, "array")
	arr[5] = {}
	arr[4] = 0
end

natives["jvml.util.ArrayList"]["remove(Ljava/lang/Object;)Z"] = function(this, e)
	local arr = getObjectField(this, "array")
	for i=1, arr[4] do
		if arr[5][i] == e then
			table.remove(arr[5], i)
			arr[4] = arr[4] - 1
			return 1
		end
	end
	return 0
end

natives["jvml.util.ArrayList"]["remove(I)Ljava/lang/Object;"] = function(this, i)
	local arr = getObjectField(this, "array")
	assert(i < arr[4] and i >= 0, "ArrayList index out of bounds")
	local e = table.remove(arr[5], i + 1)
	arr[4] = arr[4] - 1
	return e
end